packages/loopring_v3/contracts/amm/libamm/AmmJoinProcess.sol

```diff
@@ -0,0 +1,211 @@
+        AmmData.PoolJoin memory  join,
+        bytes            memory  signature
+        )
+        internal
+    {
+        S.validatePoolTransaction(
+            join.owner,
+            AmmUtil.hashPoolJoin(ctx.domainSeperator, join),
+            signature
+        );
+
+        // Check if the requirements are fulfilled
+        (bool slippageRequirementMet, uint poolAmountOut, uint96[] memory amounts) = _calculateJoinAmounts(ctx, join);
+
+        if (!slippageRequirementMet) return;
+
+        for (uint i = 0; i < ctx.size; i++) {
+            uint96 amount = amounts[i];
+
+            if (join.fromLayer2) {
+                TransferTransaction.Transfer memory transfer = ctx._block.readTransfer(ctx.txIdx++);
+                ctx.numTransactionsConsumed++;
+
+                // We do not check these fields: fromAccountID, to, amount, feeTokenID, fee
+                require(
+                    transfer.toAccountID == S.accountID &&
+                    transfer.from == join.owner &&
+                    transfer.tokenID == ctx.tokens[i].tokenID &&
+                    transfer.storageID == (signature.length > 0 ? join.storageIDs[i] : 0), // Question (brecht):is this right?
+                    "INVALID_INBOUND_TX_DATA"
+                );
+
+                // Now approve this transfer
+                transfer.validUntil = 0xffffffff;
+                bytes32 txHash = TransferTransaction.hashTx(ctx.exchangeDomainSeparator, transfer);
+                ctx.exchange.approveTransaction(join.owner, txHash);
+
+                uint96 refundAmount = 0;
+                if (!transfer.amount.isAlmostEqual(amount)) {
+                    refundAmount = transfer.amount.sub(amount);
+                }
+
+                if (refundAmount > 0) { // Process the outbound transfer
+                    transfer = ctx._block.readTransfer(ctx.txIdx++);
+                    ctx.numTransactionsConsumed++;
+
+                    // We do not check these fields: toAccountID
+                    require(
+                        transfer.fromAccountID == S.accountID &&
+                        transfer.from == address(this) &&
+                        transfer.to == join.owner &&
+                        transfer.tokenID == ctx.tokens[i].tokenID &&
+                        transfer.amount.isAlmostEqual(refundAmount) &&
+                        transfer.feeTokenID == 0 &&
+                        transfer.fee == 0 &&
+                        transfer.storageID == 0, // Question (brecht):should we check this?
+                        "INVALID_OUTBOUND_TX_DATA"
+                    );
+
+                   // Now approve this transfer
+                    transfer.validUntil = 0xffffffff;
+                    txHash = TransferTransaction.hashTx(ctx.exchangeDomainSeparator, transfer);
+                    ctx.exchange.approveTransaction(address(this), txHash);
+                    amount = amount.sub(transfer.amount);
+                }
```

Comment:
I really don't think we should do this. Rounding errors are very small, if we are worried about these extremely small rounding errors (which the operator fully controls) we can always force that the users are a disadvantage instead of the pool (but the operator can already do this, the smart contract would just enforce this). These are the same rounding errors that happen when trading, so if we are okay with them there we should be okay with them here.

Reply:
This is not because of the rounding errors Brecht.

When a user wants to join on Layer2, he will first query from the relayer how much tokens he will transfer to the AMM pool, then sign the corresponding transfers. This takes time, maybe several seconds or even a minute.  The pool is constantly being used for swapping, so when the relayer process these layer-2 join transfers, the amount that can be accepted may have changed already. So we have to have a way to refund overpaid transfers.

The layer-1 join is different - the contract only deducts tokens that have really been accepted.

Reply:
fixed